Wire tag metadata throughout the project

carterkozak opened the pull request:
This is the simpler of our available options, which leverages the
existing metadata map when spans are closed rather than replacing
the operation name string with a tuple of string and map.
Using many of our abstractions, this allows us to include *result*
data in tag values rather than only what we know up front. I've
added a simple status code tag for conjure-undertow, however
we could include all safe request headers, safe path portions,
and more. That should occur separately from the initial wiring
though.
However, many of our wrapper utility functions don't map so
nicely to this model, and take both operation name as well
as a map upon construction. It may be helpful to create a
type which encapsulates the two.

## TagRecorder details

The "low level" api for applying tags is optimized to prevent unnecessary,
and unintentional overhead both when tracing is enabled, and when it's
disabled.

Given a function with data `MyComplexObject` where we want to apply tags `foo` from `myComplexObject.getFoo()` and `bar` from `myComplexObject.getBar()`, we don't need to construct a map or even allocate a lambda:

```java
void func() {
  Tracer.fastStartSpan("loading a MyComplexObject");
  MyComplexObject object = loadMyComplexObject();
  Tracer.fastCompleteSpan(MyComplexObjectRecorder.INSTANCE, object);
}

private enum MyComplexObjectRecorder implements TagRecorder<MyComplexObject> {
    INSTANCE;

    <T> void record(TagAdapter<T> sink, T target, MyComplexObject data) {
        sink.tag(target, "foo", data.getFoo());
        sink.tag(target, "bar", data.getBar());
    }
}
```

A second `Map<String,String>` API is provided for ease of use outside of hot paths. This
is generally sufficient outside of inner-loop code:

```java
void func() {
  Tracer.fastStartSpan("loading a MyComplexObject");
  MyComplexObject object = loadMyComplexObject();
  Tracer.fastCompleteSpan(ImmutableMap.of("foo", object.getFoo(), "bar", object.getBar()));
}
```

In practice the above code delegates the call to the first API using an internal `TagRecorder<Map<String,String>>` implementation.

## Before this PR
no tags

## After this PR
==COMMIT_MSG==
Wire tag metadata throughout tracing-java
==COMMIT_MSG==

## Possible downsides?
more api surface. Not the easiest to use.

carterkozak committed:
Wire tag metadata throughout the project

This is the simpler of our available options, which leverages the
existing metadata map when spans are closed rather than replacing
the operation name string with a tuple of string and map.
Using many of our abstractions, this allows us to include *result*
data in tag values rahter than only what we know up front. I've
added a simple status code tag for conjure-undertow, however
we could include all safe request headers, safe path portions,
and more. That should occur spearately from the initial wiring
though.

carterkozak committed:
Add generated changelog entries

jkozlowski commented on tracing-undertow/src/main/java/com/palantir/tracing/undertow/UndertowTracing.java:
```diff
@@ -101,14 +103,28 @@ public void exchangeEvent(HttpServerExchange exchange, NextListener nextListener
         }
     }
 
+    private static final ImmutableMap<String, String> METADATA_200 = ImmutableMap.of("status", "200");
```
Is this some sort of standard field?

carterkozak replied:
Nope, I just figured it would be helpful to tag requests with a status code since it allows us to quickly differentiate successes from failures and demonstrate how the tagging feature works. I've updated the implementation a tad since this comment.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/CloseableTracer.java:
```diff
@@ -17,13 +17,14 @@
  * <p>Usage: try (CloseableTracer trace = CloseableTracer.start("traceName")) { [...] }
  */
-public final class CloseableTracer implements AutoCloseable {
+public class CloseableTracer implements AutoCloseable {
```
Uuuu, I wonder why this is...

carterkozak committed:
Prototype sympathetic tagging API

This API is a tad more difficult to understand than a simple
map, however it allows us to do zero work (zero allocations) in
the unsampled case, and avoid unnecessary hash tables in all
cases.

carterkozak committed:
maps

carterkozak committed:
simplify

carterkozak committed:
cleanup + apply to UndertowTracing

carterkozak committed:
a tad more

carterkozak committed:
Map+reporter api

carterkozak committed:
PECS: producer extends, consumer super

carterkozak committed:
Style

carterkozak committed:
revapi

carterkozak committed:
wiring

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -305,12 +304,18 @@ static boolean isSampled(DetachedSpan detachedSpan) {
+        private static final int NOT_COMPLETE = 0;
+        private static final int COMPLETE = 1;
+
+        private static final AtomicIntegerFieldUpdater<SampledDetachedSpan> completedUpdater =
```
Feels like a separate PR?

jkozlowski replied:
Did you run some benchmarks or something?

carterkozak replied:
This was a total driveby (but I think the only one in this PR). An attempt to reclaim some of the heap bloat from additional fields in objects.

jkozlowski commented on tracing/src/test/java/com/palantir/tracing/CloseableTracerTest.java:
```diff
@@ -53,4 +58,26 @@ public void startsAndClosesSpanWithType() {
         }
         assertThat(Tracer.getAndClearTrace().top()).isEmpty();
     }
+
```
Can you add a test for something that uses the lower level API? mostly for documentation?

carterkozak replied:
I've added a test which uses the API, but it's not very good documentation since it defines the recorder inline where in practice a single recorder would be reused (although to be fair, the inner recorder is still far more efficient than the map codepath if objects aren't being reused)

carterkozak replied:
I updated the description of this PR describing the new abstraction, hopefully that's more helpful.

carterkozak committed:
jersey status tag

carterkozak commented on tracing/src/main/java/com/palantir/tracing/CloseableTracer.java:
```diff
@@ -17,13 +17,14 @@
  * <p>Usage: try (CloseableTracer trace = CloseableTracer.start("traceName")) { [...] }
  */
-public final class CloseableTracer implements AutoCloseable {
+public class CloseableTracer implements AutoCloseable {
```
Despite not being final, the private ctor ensures that this cannot be extended.

jkozlowski replied:
Waaait how does that work: it's not private when it's in the same file? today I learned a new trick!

carterkozak replied:
heh, it's a little wonky. Our FinalClass errorprone rule has to check for this case https://github.com/palantir/gradle-baseline/blob/0f716d5ec6063b4dae06b5f761f252178aac3941/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/FinalClass.java#L100-L102

jkozlowski replied:
Actually just to set the record straight, I think I knew about this, but it always felt meh. But here it feels rather justified.

jkozlowski commented on tracing-undertow/src/main/java/com/palantir/tracing/undertow/UndertowTracing.java:
```diff
@@ -101,14 +102,34 @@ public void exchangeEvent(HttpServerExchange exchange, NextListener nextListener
+                case 204:
+                    return "204";
+            }
+            return Integer.toString(statusCode);
```
Hmm what's the impact of potentially allocating here?

jkozlowski replied:
I feel like I remember the jvm interns up to some number automatically?

carterkozak replied:
Overhead is minimal, but easy enough to avoid entirely in the most common cases, so I took that route.

carterkozak commented on tracing/src/main/java/com/palantir/tracing/DeferredTracer.java:
```diff
@@ -88,7 +93,7 @@ public DeferredTracer(Optional<String> operation) {
     }
 
-    public DeferredTracer(String operation) {
+    public DeferredTracer(String operation, Map<String, String> metadata) {
```
I opted against including a TagRecorder constructor for DeferredTracer because there's no guarantee that implementations will be serializable, which is important for spark jobs. Additionally, these are most frequently used for executors where the tags are constant across a single executor instance.

jkozlowski replied:
This is an interesting, quite subtle constraint with the Serializable thing: should we encode that somehow here? checkstyle/errorprone, or at least a comment?

carterkozak replied:
The unit test validates that objects can be serialized, but it wouldn't fail if we happened to pass serializable implementations. We wouldn't want to serialize the entire raw input in that case either... serialization is way too subtle, and I've enjoyed not having to work with it for the last year ;-)

An error-prone rule for serializable data would be nifty, I've been burned by many serialization failures! Not sure how we'd set up such a check to avoid false positives though.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/CloseableTracer.java:
```diff
@@ -46,8 +54,38 @@ public static CloseableTracer startSpan(String operation, SpanType spanType) {
+        return new TaggedCloseableTracer<>(recorder, data);
+    }
+
     @Override
```
Would it be useful to add a close that takes args? I feel like this one is maybe the most used one, so might be useful to be able to add args?

jkozlowski replied:
Hmm I guess you can also pass mutable data arg to to the constructor

carterkozak replied:
Definitely wouldn't recommend passing mutable data, heh.
Unfortunately a close method with args wouldn't be compatible with how the CloseableTracer is most often used, within a try-with-resources block. I don't want folks to give up that safety

carterkozak commented on tracing/src/main/java/com/palantir/tracing/DetachedSpan.java:
```diff
@@ -69,7 +70,33 @@ static DetachedSpan start(
+    }
+
+    @MustBeClosed
+    <T> CloseableSpan childSpan(String operationName, TagRecorder<? super T> recorder, T data, SpanType type);
```
Note I did not update the `completeAndStartChild` methods defined below (not visible in the default GH diff) because it's not obvious that the `operationName` argument would be associated with the child trace, while tags would be associated with its parent.

It's easy enough to implement that logic as needed, and produces more obvious code:
```java
 try (CloseableSpan child = detached.childSpan(operationName, type)) {
    detached.complete(<tags>);
    // work    
}
 ```

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TagRecorder.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @param <S> Stateful object type, an instance of which is passed through method calls with the {@link TagRecorder}.
+ */
+public interface TagRecorder<S> {
```
Naming is a little funny here:

Tracing has had an (unused) concept of `map<string,string>` metadata for years, so I retained the metadata name to describe the map, however here I call this a TagRecorder rather than a MetadataRecorder. I'd appreciate a second opinion. Maybe I rename all the `metadata` map parameter names to `tags`?

I bobbled between a few suffixes before using `Recorder`. I think recorder is probably fine, but `Translator` could also make sense. What do you think?

jkozlowski replied:
I think Recorder is good, maybe Translator vibes more with me, mostly because of this precedent: https://github.com/LMAX-Exchange/disruptor/blob/master/src/main/java/com/lmax/disruptor/EventTranslator.java#L27.

I would also maybe say TagAdapter -> TagSink

carterkozak replied:
I borrowed the translator idea from disruptor ;-) I'll rename to translator since I think it provides more meaning.

> I would also maybe say TagAdapter -> TagSink

Hmm, sink doesn't feel quite right because it isn't a destination, so much as a pass-through to `T target` which is the sink. Although that may not help readability....

I definitely need to add more javadoc in this interface.

jkozlowski reviewed (commented):
First batch of comments/questions from reading in the browser; I'll give it a proper read as well

carterkozak committed:
Metadata test coverage

carterkozak committed:
rename state->data

carterkozak committed:
cleaner parameter name

carterkozak committed:
fix

carterkozak commented:
Dialogue example here: https://github.com/palantir/dialogue/pull/1208

jkozlowski commented on tracing-jersey/src/main/java/com/palantir/tracing/jersey/TraceEnrichingFilter.java:
```diff
@@ -120,4 +120,13 @@ private static Observability getObservabilityFromHeader(ContainerRequestContext
+
+        @Override
+        public <T> void record(TagAdapter<T> sink, T target, ContainerResponseContext data) {
+            sink.tag(target, "status", Integer.toString(data.getStatus()));
```
Seeing the dialogue one made me think: would it not make sense to have a #tag method that takes some more common values too? like int/long/boolean? or just object?

carterkozak replied:
Hmm, I could go either way on this. I've repeated the Integer.toString piece a few times, for the same status code data. The current structure makes it obvious that everything is encoded into a string for the tracing system, where object may lead to confusion about the representation (how is ArrayList of [1,2,3] represented? not obvious that toString values aren't helpful in tracing, and may be more expensive than the tracing operation itself).

We have the flexibility to add those overloads later on using default methods if we need to.

jkozlowski replied:
Well, deep down in the rabbithole, this just adds it to a conjure type, which eventually gets shipped as jackson... Anyway, probably not worth it, but it looked like it could be a useful extension.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/CloseableTracer.java:
```diff
@@ -35,6 +36,13 @@ public static CloseableTracer startSpan(String operation) {
+    /**
+     * Opens a new {@link SpanType#LOCAL LOCAL} span for this thread's call trace, labeled with the provided operation.
+     */
+    public static CloseableTracer startSpan(String operation, Map<String, String> metadata) {
```
I know there's precedent, but shouldn't this be startLocalSpan?

carterkozak replied:
It depends. Most uses of this class are for `LOCAL` spans, so I think it's helpful to make it obvious that's the simplest choice without necessarily understanding what the span types mean. Realistically there's not much reason for CloseableTracer to support other span types at all -- server incoming requires additional data that CloseableTracer doesn't take (is_sampled, traceId, parentSpanId, etc). I suppose it could work with client outgoing spans now that we have the Tracer.maybeGetMetadata method, but that's almost exclusively used in code we own.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/DeferredTracer.java:
```diff
@@ -151,6 +166,25 @@ public void close() {
+
+        @Override
+        public void close() {
+            Tracer.fastCompleteSpan(metadata);
```
Can we extract a small method: 

```
completeAndClear(@Nullable Map<String, String> metadata) {
  Tracer.fastCompleteSpan(metadata);
  if (Tracer.hasTraceId()) {
                Tracer.getAndClearTrace();
  }
}
```

and use that here and DefaultCloseableTrace#close()?

carterkozak replied:
I think this method can be simplified to avoid the `getAndClear` branch when an original trace is present:
```java
Tracer.fastCompleteSpan(metadata);
if (originalTrace.isPresent()) {
    Tracer.setTrace(originalTrace.get());
} else if (Tracer.hasTraceId()) {
    Tracer.getAndClearTrace();
}
```

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -453,29 +499,30 @@ static void fastDiscardSpan() {
-     * @deprecated Use {@link #fastCompleteSpan()}
      */
-    @Deprecated
     public static void fastCompleteSpan(Map<String, String> metadata) {
```
@Nullable right? you've designed this to be a no-op if the metadata is null?

carterkozak replied:
I don't want to support nulls in the API/contract, but I prefer for observability code to be tolerant to null inputs. I think adding the `recorder.isEmpty(data)` check in the Span.Builder adapter would allow this to accept nulls, otherwise the Span object will NPE.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/DeferredTracer.java:
```diff
@@ -88,7 +93,7 @@ public DeferredTracer(Optional<String> operation) {
     }
 
-    public DeferredTracer(String operation) {
+    public DeferredTracer(String operation, Map<String, String> metadata) {
```
Hear me out: shouldn't this be ```Map<String, Object>``? If you're being perf-conscious, a boxed Integer is surely better than String representation or that?

carterkozak replied:
Hmm, boxing an integer outside of the default -128 -> 127 (my memory on exact ranges here is a bit fuzzy, still need coffee) default `int`->`Integer` cache isn't substantially better than int -> String, but valhalla will change that one day! I can make a similar argument to the overloads on TagAdapter about subtlety of object->string conversions and unexpected tag value data. I think it's safe to design around the assumption that tag values will always be strings.

This tracer type is generally used with static operation name/tag inputs, for example from wrapped executors where the name is created on initialization and passed to each deferredtracer.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/DetachedSpan.java:
```diff
@@ -69,7 +70,33 @@ static DetachedSpan start(
      */
     @MustBeClosed
-    CloseableSpan childSpan(String operationName, SpanType type);
+    default CloseableSpan childSpan(String operationName, SpanType type) {
```
Call me old-fashioned, but these default methods scare me just a little bit. And the amount of them

carterkozak replied:
Ya, it's a balancing act.

jkozlowski commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -525,17 +572,31 @@ private static void notifyObservers(Span span) {
-                .putAllMetadata(metadata)
-                .build();
+                .durationNanoSeconds(System.nanoTime() - openSpan.getStartClockNanoSeconds());
+        tag.record(SpanBuilderTagAdapter.INSTANCE, builder, state);
```
So this is quite hard to validate and probably maintain (if you make the callers do this), but *I feel* like the contract you have with TagRecorder is that you should call #isEmpty()?

```
if (!tag.isEmpty(state)) {
            tag.record(SpanBuilderTagAdapter.INSTANCE, builder, state);    
        }
```

carterkozak replied:
You're right, we might as well do the empty check here.
If it is empty, it shouldn't cost anything to call `tag.record`. The `isEmpty` method allows us to quickly validate whether or not we need to keep track of the instance in some cases.

jkozlowski reviewed (commented):
I think I'm happy to approve this, I've stared at it long enough now. I am not going to push for it, because I've done the review now, but my preference would be to actually split out into a bunch of PRs:

1. Perf things (like moving the completed field to be AtomicUpdater). Easy and self-contained, and you could probably push a benchmark as well!
2. Driveby's: like extracting little named-classes in preference of anonymous lambdas and whatnot.
3. Push new API for the tag stuff, impl does nothing.
4. Push impl, with the tests.

The whole change is quite subtle, and whilst I think I covered everything I wanted, it is still possible we both missed something.

carterkozak committed:
rename recorder->translator

carterkozak committed:
simplify deferredtracer

carterkozak committed:
update revapi

carterkozak commented:
> Perf things (like moving the completed field to be AtomicUpdater). Easy and self-contained, and you could probably push a benchmark as well!

Ya, this would have been better as another PR. Unfortunately stacking in GH isn't as easy as I'd like, and I got lazy pushing through the AtomicIntegerFieldUpdater change.

> Driveby's: like extracting little named-classes in preference of anonymous lambdas and whatnot.

I think all of these were required to ferry around the new translator/data objects. I may have added something that I've forgotten about though!

> Push new API for the tag stuff, impl does nothing.
> Push impl, with the tests.

I wouldn't want to merge the API with a no-op impl because it allows folks to start migrating to it, but lose data in practice. I agree that it's best to review API and impl changes separately. It's unfortunate that this project does not separate API from implementation at all, which I'd like to resolve in a future tracing v2 library, but who knows when we'll get to that :-/

carterkozak committed:
additional javadoc

svc-autorelease commented:
Released [4.14.0](https://github.com/palantir/tracing-java/releases/tag/4.14.0)
